removed the delay for spi_comm

neelrparikh opened the pull request:
I think this should work. I wasn't too sure on the pin number used by boot0, I have used GPIO4 on the ESP and B0 on the ST

geohot commented on board/main.c:
```diff
@@ -740,10 +740,22 @@ void handle_spi(uint8_t *data, int len) {
     case 0:
+	  // handshake with esp
+	  // Set boot0 to output and pull high
+	  GPIOB->MODER &= ~(0x3);
```
Don't use a constant here

geohot commented on boardesp/proxy.c:
```diff
@@ -51,10 +51,14 @@ static int ICACHE_FLASH_ATTR __spi_comm(char *dat, int len, uint32_t *recvData,
+  // set boot0 as input
+  gpio_output_set(0,0,0,(1 << 4));
+  //wait for boot0 to be set to LOW by ST, counter to avoid infinite loop
+  for(int i = 0;(gpio_input_get() & (1 << 4)) && i < 100; i++) {
```
What happens if this counter expires and data still isn't ready?

neelrparikh replied:
The counter makes it wait for 1ms, same as before. If the data isn't ready it will be an issue. The only other way I can think of would be to use an infinite loop and always poll but that would crash the system if spi fails. Even using a ack would require a counter to avoid an infinite loop. Any suggestions?

geohot commented on board/main.c:
```diff
@@ -757,6 +769,9 @@ void handle_spi(uint8_t *data, int len) {
       break;
   }
   spi_tx_dma(spi_tx_buf, 0x44);
```
Is this the only place we use SPI?

neelrparikh replied:
Yes, this is the only place where we need to receive data over SPI, we transmit data in spi_flasher.h:122 but there's no need for this there as we are no waiting for receiving data to be ready.

geohot replied:
Hmm, but does the ESP know this? Won't it still wait?

neelrparikh replied:
Let me just go through it again to be 100% sure

geohot commented on board/main.c:
```diff
@@ -740,10 +740,22 @@ void handle_spi(uint8_t *data, int len) {
+	  // Set boot0 to output and pull high
+	  GPIOB->MODER &= ~(0x3);
+	  GPIOB->MODER |= GPIO_MODER_MODER0_0;
+	  GPIOB->ODR |= (1); 
```
Why is 1 in parens?

neelrparikh commented:
Done, sorry about that I got messy with the constants. Let me know if we can make some changes for the counter, can't think of anything right now (worst case we can increase the delay).

geohot commented:
lol yea, i'd just increase the delay to 10ms.

are you 100% sure you don't need this on the spi_flasher?

neelrparikh commented:
I don't think this method will work, I just realized spi_flasher uses boot0 while it is using spi. I added the ack code to spi_flasher but doubt it will work. Let me think over it and post changes tomorrow.
For now I increased the delay and added an os_print may be later we can add some kind of error message?

geohot commented:
it helps to understand what we use the boot pin for, it's just a gpio connecting the processors.

also, the library to interact with this is also open source in lib/panda.py

neelrparikh commented:
I think I got a work around for this, when accessing spi_flasher the spi has not yet been initialized so we can use the cs pin for handshaking to indicate that it has entered the flasher, after which we initialize the spi and use the boot0 pin for handshaking. Do you think it will work? I think it should.

geohot commented:
maybe, write it and i'll try.

neelrparikh commented:
okay i am pushing the code

geohot commented on boardesp/webserver.c:
```diff
@@ -62,6 +62,13 @@ void st_set_boot_mode(int boot_mode) {
+	// use cs as a handshake for spi_flasher
+	// set cs as input and wait for it to go high
+	gpio_output_set(0, 0, 0, (1 << 5));
+	//TODO: remove infinite loop
```
Not okay.

neelrparikh replied:
I can remove the infinite loop but did not know the delay to keep for the ST to start up would a worst case of 10ms be good?

geohot commented on board/spi_flasher.h:
```diff
@@ -53,6 +62,12 @@ void spi_flasher() {
     lastval = val;
 
     if (rcv && (DMA2->LISR & DMA_LISR_TCIF2)) {
+	  // handshake with esp
```
Are these tabs!?

neelrparikh replied:
Ahh! my ide inserted tabs I'll get rid of them

geohot commented on board/spi_flasher.h:
```diff
@@ -53,6 +62,12 @@ void spi_flasher() {
+	  // Set boot0 to output and pull high
+	  GPIOB->MODER &= ~(GPIO_MODER_MODER0);
+	  GPIOB->MODER |= GPIO_MODER_MODER0_0;
+	  GPIOB->ODR |= GPIO_ODR_OD0; 
```
This doesn't build

neelrparikh replied:
Can't figure why this is not building, I will try building and may be run it on an ESP I have.

geohot commented on boardesp/proxy.c:
```diff
@@ -51,10 +51,17 @@ static int ICACHE_FLASH_ATTR __spi_comm(char *dat, int len, uint32_t *recvData,
+  }
+  // print error if handshake not successful 
+  if(gpio_input_get() & (1 << 4))
+	os_print("ERROR: SPI receive failed\n");
```
What is os_print? Did you build this?

neelrparikh replied:
It's suppose to be os_printf sorry about that.

neelrparikh commented:
Sorry about the build errors, I was not able to compile in my environment. I managed to compile the files and removed all the build errors. Thanks.

geohot commented:
Good enough for a microinternship. We were in a rush to get pandas out and I ended up writing a variant of this, so I won't actually merge this. (you can read, it's committed). Viv will be in touch with microinternship details.

neelrparikh commented:
Thank you, sorry for the delay I was stuck in school today. I will have a look on what you have done! Sounds good thanks again looking forward to meeting you.
